genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -0,0 +1,363 @@
+                        fileTransfer.completeTransfer(true, new InterruptedException("Shutting down"));
+                    }
+                } catch (NoSuchElementException | ConcurrentModificationException e) {
+                    // Swallow. Not unexpected, collection and state may change.
+                }
+            }
+        }
+    }
+
+    private synchronized void pushManifest() {
+        if (started.get()) {
+            final AgentManifestMessage jobFileManifest;
+            try {
+                jobFileManifest = manifestProtoConverter.manifestToProtoMessage(
+                    this.jobId,
+                    new JobDirectoryManifest(this.jobDirectoryPath, false)
+                );
+            } catch (final IOException e) {
+                log.error("Failed to construct manifest", e);
+                return;
+            } catch (GenieConversionException e) {
+                log.error("Failed to serialize manifest", e);
+                return;
+            }
+
+            if (this.controlStreamObserver == null) {
+                this.controlStreamObserver = fileStreamServiceStub.sync(this.responseObserver);
+                if (this.controlStreamObserver instanceof ClientCallStreamObserver) {
+                    ((ClientCallStreamObserver) this.controlStreamObserver).setMessageCompression(ENABLE_COMPRESSION);
+                }
+            }
+
+            this.controlStreamObserver.onNext(jobFileManifest);
+        }
+    }
+
+    private void handleControlStreamError(final Throwable t) {
+        log.warn("Manifest stream error: {}", t.getMessage(), t);
+        this.trigger.reset();
+        this.discardCurrentStream(false);
+    }
+
+    private void handleControlStreamCompletion() {
+        log.debug("Manifest stream completed");
+        this.discardCurrentStream(false);
+    }
+
+    private synchronized void discardCurrentStream(final boolean sendStreamCompletion) {
+        if (this.controlStreamObserver != null) {
+            if (sendStreamCompletion) {
+                this.controlStreamObserver.onCompleted();
+            }
+            this.controlStreamObserver = null;
+        }
+    }
+
+    private synchronized void handleFileRequest(
+        final String streamId,
+        final String relativePath,
+        final int startOffset,
+        final int endOffset
+    ) {
+        log.info(
+            "Server is requesting file {} (range: [{}, {}, streamId: {})",
+            relativePath,
+            startOffset,
+            endOffset,
+            streamId
+        );
+
+        if (!this.started.get()) {
+            log.warn("Ignoring file request, service shutting down");
+            return;
+        }
+
+        final Path absolutePath = this.jobDirectoryPath.resolve(relativePath);
+
+        if (!Files.exists(absolutePath)) {
+            log.warn("Ignoring request for a file that does not exist: {}", absolutePath);
```

Comment:
Will this somehow propagate a 404?

Reply:
In theory, this code branch should never hit because the server consults the manifest before making a request (i.e. it won't request a file that doesn't exist).

There's a slim chance the file was deleted by the time the request arrives on the Agent, and this code will hit.
This will result in a timeout on the server that closes the inputStream on the resource.